### PR TITLE
runtime-rs: add driver option

### PR DESCRIPTION
**Pull request by Tim-0731-Hzt**
add driver option when handle linux devices

Fixes:#7539

### CODE REVIEW

**Review by Apokleos — APPROVED** [review]
LGTM!

**Review by cmaf — COMMENTED** [review]
Thanks @Tim-0731-Hzt. This solves the previous error but now I am seeing a new one with cloud hypervisor:

ctr: failed to create shim task: Others("failed to handler message create container\n\nCaused by:\n    0: create\n    1: agent create container\n    2: rpc status: Status { code: INTERNAL, message: \"cannot parse integer from empty string\", details: [], special_fields: SpecialFields { unknown_fields: UnknownFields { fields: None }, cached_size: CachedSize { size: 0 } } }"): unknown

If this is specific to cloud hypervisor I will debug. I have not tested with dragonball.

**Comment by Apokleos** [comment]
> 
>
Hi @cmaf ,is there some logs about the error ?

**Comment by cmaf** [comment]
@Apokleos this was directly before DeleteProcess:

```
Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"receive createcontainer, spec: Spec { version: \\\"1.0.2-dev\\\", process: Some(Process { terminal: true, console_size: None, user: User { uid: 0, gid: 0, additional_gids: [10], username: \\\"\\\" }, args: [\\\"sh\\\"], env: [\\\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\\\", \\\"TERM=xterm\\\"], cwd: \\\"/\\\", capabilities: Some(LinuxCapabilities { bounding: [\\\"CAP_CHOWN\\\", \\\"CAP_DAC_OVERRIDE\\\", \\\"CAP_FSETID\\\", \\\"CAP_FOWNER\\\", \\\"CAP_MKNOD\\\", \\\"CAP_NET_RAW\\\", \\\"CAP_SETGID\\\", \\\"CAP_SETUID\\\", \\\"CAP_SETFCAP\\\", \\\"CAP_SETPCAP\\\"time="2023-08-07T09:01:10.893928265-07:00" level=debug msg="failed to delete task" error="rpc error: code = Unknown desc = Others(\"failed to handler message handler request\\n\\nCaused by:\\n    0: do delete\\n    1: failed to find container test_container_972\")" id=test_container_972
Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: , \\\"CAP_NET_BIND_SERVICE\\\", \\\"CAP_SYS_CHROOT\\\", \\\"CAP_KILL\\\", \\\"CAP_AUDIT_WRITE\\\"], effective: [\\\"CAP_CHOWN\\\", \\\"CAP_DAC_OVERRIDE\\\", \\\"CAP_FSETID\\\", \\\"CAP_FOWNER\\\", \\\"CAP_MKNOD\\\", \\\"CAP_NET_RAW\\\", \\\"CAP_SETGID\\\", \\\"CAP_SETUID\\\", \\\"CAP_SETFCAP\\\", \\\"CAP_SETPCAP\\\", \\\"CAP_NET_BIND_SERVICE\\\", \\\"CAP_SYS_CHROOT\\\", \\\"CAP_KILL\\\", \\\"CAP_AUDIT_WRITE\\\"], inheritable: [], permitted: [\\\"CAP_CHOWN\\\", \\\"CAP_DAC_OVERRIDE\\\", \\\"CAP_FSETID\\\", \\\"CAP_FOWNER\\\", \\\"CAP_MKNOD\\\", \\\"CAP_NET_RAW\\\", \\\"CAP_SETGID\\\", \\\"CAP_SETUID\\\", \\\"CAP_SETFCAP\\\", \\\"CAP_SETPCAP\\\", \\\"CAP_NET_BIND_SERVICE\\\", \\\"CAP_SYS_CHROOT\\\", \\\"CAP_KILL\\\", \\\"CAP_AUDIT_WRITE\\\"], ambient: [] }), rlimits: [PosixRlimit { type: \\\"RLIMIT_NOFILE\\\", hard: 1024, soft: 1024 }], no_new_privileges: true, apparmor_profile: \\\"\\\", oom_score_adj: Some(0), selinux_label: \\\"\\\" }), root: Some(Root { path: \\\"/run/kata-containers/shared/containers/passthrough/test_container_972/rootfs\\\", readonly: false }), hostname: \\\"\\\", mounts: [Mount { destination: \\\"/proc\\\", type: \\\"proc\\\", source: \\\"proc\\\", options: [\\\"nosuid\\\", \\\"noexec\\\", \\\"nodev\\\"] }, Mount { destination: \\\"/dev\\\", type: \\\"tmpfs\\\", source: \\\"tmpfs\\\", options: [\\\"nosuid\\\", \\\"strictatime\\\", \\\"mode=755\\\", \\\"size=65536k\\\"] }, Mount { destination: \\\"/dev/pts\\\", type: \\\"devpts\\\", source: \\\"devpts\\\", options: [\\\"nosuid\\\", \\\"noexec\\\", \\\"newinstance\\\", \\\"ptmxmode=0666\\\", \\\"mode=0620\\\", \\\"gid=5\\\"] }, Mount { destination: \\\"/dev/shm\\\", type: \\\"bind\\\", source: \\\"/run/kata-containers/sandbox/shm\\\", options: [\\\"rbind\\\"] }, Mount { destination: \\\"/dev/mqueue\\\", type: \\\"mqueue\\\", source: \\\"mqueue\\\", options: [\\\"nosuid\\\", \\\"noexec\\\", \\\"nodev\\\"] }, Mount { destination: \\\"/sys\\\", type: \\\"sysfs\\\", source: \\\"sysfs\\\", options: [\\\"nosuid\\\", \\\"noexec\\\", \\\"nodev\\\", \\\"ro\\\"] }, Mount { destination: \\\"/run\\\", type: \\\"tmpfs\\\", source: \\\"tmpfs\\\", options: [\\\"nosuid\\\", \\\"strictatime\\\", \\\"mode=755\\\", \\\"size=65536k\\\"] }], hooks: None, annotations: {}, linux: Some(Linux { uid_mappings: [], gid_mappings: [], sysctl: {}, resources: Some(LinuxResources { devices: [LinuxDeviceCgroup { allow: false, type: \\\"\\\", major: Some(0), minor: Some(0), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(3), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(8), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(7), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(5), minor: Some(0), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(5), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(9), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(5), minor: Some(1), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(136), minor: Some(0), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(5), minor: Some(2), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(10), minor: Some(200), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"b\\\", major: Some(7), minor: Some(19), access: \\\"rwm\\\" }], memory: None, cpu: Some(LinuxCpu { shares: Some(1024), quota: Some(0), period: Some(0), realtime_runtime: Some(0), realtime_period: Some(0), cpus: \\\"\\\", mems: \\\"\\\" }), pids: None, block_io: None, hugepage_limits: [], network: None, rdma: {} }), cgroups_path: \\\"/default/test_container_972\\\", namespaces: [LinuxNamespace { type: \\\"ipc\\\", path: \\\"\\\" }, LinuxNamespace { type: \\\"uts\\\", path: \\\"\\\" }, LinuxNamespace { type: \\\"mount\\\", path: \\\"\\\" }], devices: [LinuxDevice { path: \\\"/dev/loop19\\\", type: \\\"b\\\", major: 7, minor: 19, file_mode: Some(432), uid: Some(0), gid: Some(6) }], seccomp: None, rootfs_propagation: \\\"\\\", masked_paths: [\\\"/proc/acpi\\\", \\\"/proc/asound\\\", \\\"/proc/kcore\\\", \\\"/proc/keys\\\", \\\"/proc/latency_stats\\\", \\\"/proc/timer_list\\\", \\\"/proc/timer_stats\\\", \\\"/proc/sched_debug\\\", \\\"/sys/firmware\\\", \\\"/proc/scsi\\\"], readonly_paths: [\\\"/proc/bus\\\", \\\"/proc/fs\\\", \\\"/proc/irq\\\", \\\"/proc/sys\\\", \\\"/proc/sysrq-trigger\\\"], mount_label: \\\"\\\", intel_rdt: None }), solaris: None, windows: None, vm: None }\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.842782541Z\",\"pid\":\"108\",\"subsystem\":\"rpc\",\"version\":\"0.1.0\",\"source\":\"agent\",\"name\":\"kata-agent\"}","level":"INFO","ts":"2023-08-07T16:01:10.894721534Z","name":"kata-runtime","pid":"80459","subsystem":"agent","version":"0.1.0","source":"test_container_972"}
Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"receive createcontainer, storages: [Storage { driver: \\\"ephemeral\\\", driver_options: [], source: \\\"shm\\\", fstype: \\\"tmpfs\\\", options: [\\\"noexec\\\", \\\"nosuid\\\", \\\"nodev\\\", \\\"mode=1777\\\", \\\"size=67108864\\\"], mount_point: \\\"/run/kata-containers/sandbox/shm\\\", fs_group: MessageField(None), special_fields: SpecialFields { unknown_fields: UnknownFields { fields: None }, cached_size: CachedSize { size: 0 } } }]\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.842839407Z\",\"version\":\"0.1.0\",\"subsystem\":\"rpc\",\"pid\":\"108\",\"source\":\"agent\",\"name\":\"kata-agent\"}","level":"INFO","ts":"2023-08-07T16:01:10.894886753Z","subsystem":"agent","name":"kata-runtime","pid":"80459","source":"test_container_972","version":"0.1.0"}
Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"device-id: /dev/vda, device-type: blk, device-vm-path: /dev/vda, device-container-path: /dev/loop19, device-options: []\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.842877228Z\",\"version\":\"0.1.0\",\"name\":\"kata-agent\",\"subsystem\":\"device\",\"source\":\"agent\",\"pid\":\"108\"}","level":"INFO","ts":"2023-08-07T16:01:10.894998049Z","source":"test_container_972","pid":"80459","subsystem":"agent","version":"0.1.0","name":"kata-runtime"}
Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"TEST_agent virtio_blk_device_handler device_id: /dev/vda\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.842915015Z\",\"version\":\"0.1.0\",\"pid\":\"108\",\"name\":\"kata-agent\",\"subsystem\":\"device\",\"source\":\"agent\"}","level":"INFO","ts":"2023-08-07T16:01:10.895108247Z","pid":"80459","subsystem":"agent","version":"0.1.0","source":"test_container_972","name":"kata-runtime"}
Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"got uevent message\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.84308667Z\",\"subsystem\":\"uevent\",\"source\":\"agent\",\"name\":\"kata-agent\",\"pid\":\"108\",\"version\":\"0.1.0\",\"event\":\"Uevent { action: \\\"add\\\", devpath: \\\"/devices/virtual/bdi/254:0\\\", devname: \\\"\\\", subsystem: \\\"bdi\\\", seqnum: \\\"629\\\", interface: \\\"\\\" }\"}","level":"INFO","ts":"2023-08-07T16:01:10.895225459Z","subsystem":"agent","pid":"80459","name":"kata-runtime","version":"0.1.0","source":"test_container_972"}
```

**Comment by Apokleos** [comment]
> @Apokleos this was directly before DeleteProcess:
> 
> ```
> Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"receive createcontainer, spec: Spec { version: \\\"1.0.2-dev\\\", process: Some(Process { terminal: true, console_size: None, user: User { uid: 0, gid: 0, additional_gids: [10], username: \\\"\\\" }, args: [\\\"sh\\\"], env: [\\\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\\\", \\\"TERM=xterm\\\"], cwd: \\\"/\\\", capabilities: Some(LinuxCapabilities { bounding: [\\\"CAP_CHOWN\\\", \\\"CAP_DAC_OVERRIDE\\\", \\\"CAP_FSETID\\\", \\\"CAP_FOWNER\\\", \\\"CAP_MKNOD\\\", \\\"CAP_NET_RAW\\\", \\\"CAP_SETGID\\\", \\\"CAP_SETUID\\\", \\\"CAP_SETFCAP\\\", \\\"CAP_SETPCAP\\\"time="2023-08-07T09:01:10.893928265-07:00" level=debug msg="failed to delete task" error="rpc error: code = Unknown desc = Others(\"failed to handler message handler request\\n\\nCaused by:\\n    0: do delete\\n    1: failed to find container test_container_972\")" id=test_container_972
> Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: , \\\"CAP_NET_BIND_SERVICE\\\", \\\"CAP_SYS_CHROOT\\\", \\\"CAP_KILL\\\", \\\"CAP_AUDIT_WRITE\\\"], effective: [\\\"CAP_CHOWN\\\", \\\"CAP_DAC_OVERRIDE\\\", \\\"CAP_FSETID\\\", \\\"CAP_FOWNER\\\", \\\"CAP_MKNOD\\\", \\\"CAP_NET_RAW\\\", \\\"CAP_SETGID\\\", \\\"CAP_SETUID\\\", \\\"CAP_SETFCAP\\\", \\\"CAP_SETPCAP\\\", \\\"CAP_NET_BIND_SERVICE\\\", \\\"CAP_SYS_CHROOT\\\", \\\"CAP_KILL\\\", \\\"CAP_AUDIT_WRITE\\\"], inheritable: [], permitted: [\\\"CAP_CHOWN\\\", \\\"CAP_DAC_OVERRIDE\\\", \\\"CAP_FSETID\\\", \\\"CAP_FOWNER\\\", \\\"CAP_MKNOD\\\", \\\"CAP_NET_RAW\\\", \\\"CAP_SETGID\\\", \\\"CAP_SETUID\\\", \\\"CAP_SETFCAP\\\", \\\"CAP_SETPCAP\\\", \\\"CAP_NET_BIND_SERVICE\\\", \\\"CAP_SYS_CHROOT\\\", \\\"CAP_KILL\\\", \\\"CAP_AUDIT_WRITE\\\"], ambient: [] }), rlimits: [PosixRlimit { type: \\\"RLIMIT_NOFILE\\\", hard: 1024, soft: 1024 }], no_new_privileges: true, apparmor_profile: \\\"\\\", oom_score_adj: Some(0), selinux_label: \\\"\\\" }), root: Some(Root { path: \\\"/run/kata-containers/shared/containers/passthrough/test_container_972/rootfs\\\", readonly: false }), hostname: \\\"\\\", mounts: [Mount { destination: \\\"/proc\\\", type: \\\"proc\\\", source: \\\"proc\\\", options: [\\\"nosuid\\\", \\\"noexec\\\", \\\"nodev\\\"] }, Mount { destination: \\\"/dev\\\", type: \\\"tmpfs\\\", source: \\\"tmpfs\\\", options: [\\\"nosuid\\\", \\\"strictatime\\\", \\\"mode=755\\\", \\\"size=65536k\\\"] }, Mount { destination: \\\"/dev/pts\\\", type: \\\"devpts\\\", source: \\\"devpts\\\", options: [\\\"nosuid\\\", \\\"noexec\\\", \\\"newinstance\\\", \\\"ptmxmode=0666\\\", \\\"mode=0620\\\", \\\"gid=5\\\"] }, Mount { destination: \\\"/dev/shm\\\", type: \\\"bind\\\", source: \\\"/run/kata-containers/sandbox/shm\\\", options: [\\\"rbind\\\"] }, Mount { destination: \\\"/dev/mqueue\\\", type: \\\"mqueue\\\", source: \\\"mqueue\\\", options: [\\\"nosuid\\\", \\\"noexec\\\", \\\"nodev\\\"] }, Mount { destination: \\\"/sys\\\", type: \\\"sysfs\\\", source: \\\"sysfs\\\", options: [\\\"nosuid\\\", \\\"noexec\\\", \\\"nodev\\\", \\\"ro\\\"] }, Mount { destination: \\\"/run\\\", type: \\\"tmpfs\\\", source: \\\"tmpfs\\\", options: [\\\"nosuid\\\", \\\"strictatime\\\", \\\"mode=755\\\", \\\"size=65536k\\\"] }], hooks: None, annotations: {}, linux: Some(Linux { uid_mappings: [], gid_mappings: [], sysctl: {}, resources: Some(LinuxResources { devices: [LinuxDeviceCgroup { allow: false, type: \\\"\\\", major: Some(0), minor: Some(0), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(3), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(8), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(7), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(5), minor: Some(0), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(5), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(1), minor: Some(9), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(5), minor: Some(1), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(136), minor: Some(0), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(5), minor: Some(2), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"c\\\", major: Some(10), minor: Some(200), access: \\\"rwm\\\" }, LinuxDeviceCgroup { allow: true, type: \\\"b\\\", major: Some(7), minor: Some(19), access: \\\"rwm\\\" }], memory: None, cpu: Some(LinuxCpu { shares: Some(1024), quota: Some(0), period: Some(0), realtime_runtime: Some(0), realtime_period: Some(0), cpus: \\\"\\\", mems: \\\"\\\" }), pids: None, block_io: None, hugepage_limits: [], network: None, rdma: {} }), cgroups_path: \\\"/default/test_container_972\\\", namespaces: [LinuxNamespace { type: \\\"ipc\\\", path: \\\"\\\" }, LinuxNamespace { type: \\\"uts\\\", path: \\\"\\\" }, LinuxNamespace { type: \\\"mount\\\", path: \\\"\\\" }], devices: [LinuxDevice { path: \\\"/dev/loop19\\\", type: \\\"b\\\", major: 7, minor: 19, file_mode: Some(432), uid: Some(0), gid: Some(6) }], seccomp: None, rootfs_propagation: \\\"\\\", masked_paths: [\\\"/proc/acpi\\\", \\\"/proc/asound\\\", \\\"/proc/kcore\\\", \\\"/proc/keys\\\", \\\"/proc/latency_stats\\\", \\\"/proc/timer_list\\\", \\\"/proc/timer_stats\\\", \\\"/proc/sched_debug\\\", \\\"/sys/firmware\\\", \\\"/proc/scsi\\\"], readonly_paths: [\\\"/proc/bus\\\", \\\"/proc/fs\\\", \\\"/proc/irq\\\", \\\"/proc/sys\\\", \\\"/proc/sysrq-trigger\\\"], mount_label: \\\"\\\", intel_rdt: None }), solaris: None, windows: None, vm: None }\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.842782541Z\",\"pid\":\"108\",\"subsystem\":\"rpc\",\"version\":\"0.1.0\",\"source\":\"agent\",\"name\":\"kata-agent\"}","level":"INFO","ts":"2023-08-07T16:01:10.894721534Z","name":"kata-runtime","pid":"80459","subsystem":"agent","version":"0.1.0","source":"test_container_972"}
> Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"receive createcontainer, storages: [Storage { driver: \\\"ephemeral\\\", driver_options: [], source: \\\"shm\\\", fstype: \\\"tmpfs\\\", options: [\\\"noexec\\\", \\\"nosuid\\\", \\\"nodev\\\", \\\"mode=1777\\\", \\\"size=67108864\\\"], mount_point: \\\"/run/kata-containers/sandbox/shm\\\", fs_group: MessageField(None), special_fields: SpecialFields { unknown_fields: UnknownFields { fields: None }, cached_size: CachedSize { size: 0 } } }]\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.842839407Z\",\"version\":\"0.1.0\",\"subsystem\":\"rpc\",\"pid\":\"108\",\"source\":\"agent\",\"name\":\"kata-agent\"}","level":"INFO","ts":"2023-08-07T16:01:10.894886753Z","subsystem":"agent","name":"kata-runtime","pid":"80459","source":"test_container_972","version":"0.1.0"}
> Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"device-id: /dev/vda, device-type: blk, device-vm-path: /dev/vda, device-container-path: /dev/loop19, device-options: []\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.842877228Z\",\"version\":\"0.1.0\",\"name\":\"kata-agent\",\"subsystem\":\"device\",\"source\":\"agent\",\"pid\":\"108\"}","level":"INFO","ts":"2023-08-07T16:01:10.894998049Z","source":"test_container_972","pid":"80459","subsystem":"agent","version":"0.1.0","name":"kata-runtime"}
> Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"TEST_agent virtio_blk_device_handler device_id: /dev/vda\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.842915015Z\",\"version\":\"0.1.0\",\"pid\":\"108\",\"name\":\"kata-agent\",\"subsystem\":\"device\",\"source\":\"agent\"}","level":"INFO","ts":"2023-08-07T16:01:10.895108247Z","pid":"80459","subsystem":"agent","version":"0.1.0","source":"test_container_972","name":"kata-runtime"}
> Aug 07 09:01:10 cmaf-M4HM87P-00 containerd[4831]: {"msg":"{\"msg\":\"got uevent message\",\"level\":\"INFO\",\"ts\":\"2023-08-07T16:01:10.84308667Z\",\"subsystem\":\"uevent\",\"source\":\"agent\",\"name\":\"kata-agent\",\"pid\":\"108\",\"version\":\"0.1.0\",\"event\":\"Uevent { action: \\\"add\\\", devpath: \\\"/devices/virtual/bdi/254:0\\\", devname: \\\"\\\", subsystem: \\\"bdi\\\", seqnum: \\\"629\\\", interface: \\\"\\\" }\"}","level":"INFO","ts":"2023-08-07T16:01:10.895225459Z","subsystem":"agent","pid":"80459","name":"kata-runtime","version":"0.1.0","source":"test_container_972"}
> ```
>
Hi @cmaf , 
Could you help to hack it that fill the field `id` with a correct index replacing the  `device.config.virt_path.clone()` at [line](https://github.com/kata-containers/kata-containers/blob/5a20d8dcaf5b4c0bd61aa1495ad56e293d13de37/src/runtime-rs/crates/resource/src/manager_inner.rs#L294) ?

When the block_device_driver is "virtio-blk-pci", the agent device needs a device.Id with the PCI path given, but now `the device-id: /dev/vda`, so I want to confirm whether the device-id causes the error.

**Comment by justxuewei** [comment]
> If this is specific to cloud hypervisor I will debug. I have not tested with dragonball.

Hi @cmaf, I've test on my environment with this patch. Everything goes smoothly. FYI, the VMM I used is the Dragonball, the block_device_driver is `virtio-blk-mmio`.

**Review by justxuewei — APPROVED** [review]
Lgtm, thanks!

**Review by bergwolf — APPROVED** [review]
lgtm! Not a blocker but it would be nice if you can add a test case for the block device driver setting.

**Review by lifupan — APPROVED** [review]

**Comment by cmaf** [comment]
> Hi @cmaf , Could you help hack and then test it that fill the field `id` with a correct index replacing the `device.config.virt_path.clone()` at [line](https://github.com/kata-containers/kata-containers/blob/5a20d8dcaf5b4c0bd61aa1495ad56e293d13de37/src/runtime-rs/crates/resource/src/manager_inner.rs#L294) ?
> 
> When the block_device_driver is "virtio-blk-pci", the agent device needs a device.Id with the PCI path given, but now `the device-id: /dev/vda`, so I want to confirm whether the device-id causes the error.

@Apokleos it actually does work with CH with the correct index hard coded

**Comment by Apokleos** [comment]
> > Hi @cmaf , Could you help hack and then test it that fill the field `id` with a correct index replacing the `device.config.virt_path.clone()` at [line](https://github.com/kata-containers/kata-containers/blob/5a20d8dcaf5b4c0bd61aa1495ad56e293d13de37/src/runtime-rs/crates/resource/src/manager_inner.rs#L294) ?
> > When the block_device_driver is "virtio-blk-pci", the agent device needs a device.Id with the PCI path given, but now `the device-id: /dev/vda`, so I want to confirm whether the device-id causes the error.
> 
> @Apokleos it actually does work with CH with the correct index hard coded
>
This is caused by the fact that the pci address of the virtio-blk-pci device is not correctly specified when constructing the agent device. I am currently working on pci device topology to address it , please check my issue here: https://github.com/kata-containers/kata-containers/issues/7218

Can we merge the changes first, and then discuss the PCI address in issue #7218, if the fix for "runtime-rs: add driver option" is okay ?

**Comment by cmaf** [comment]
@Apokleos @Tim-0731-Hzt yes, I'm ok with merging this and resolving the rest in https://github.com/kata-containers/kata-containers/issues/7218. Thanks!

**Review by jodh-intel — APPROVED** [review]
Thanks @Tim-0731-Hzt.

**Comment by Tim-0731-Hzt** [comment]
thanks for your review, I will add some unit test for this before merge this PR

File: src/runtime-rs/crates/hypervisor/src/device/device_manager.rs [hunk 1]
```diff
@@ -470,3 +470,69 @@ pub async fn do_handle_device(
...
+            .clone()
+            .read()
+            .await
+            .devices
+            .get(&device_id)
+            .unwrap()
+            .lock()
+            .await
+            .get_device_info()
+            .await
```
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/device/device_manager.rs:531]: Here, IMO, you'd better do assert it before using `.unwrap()` in L520 and L527.

Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/hypervisor/src/device/device_manager.rs]: thanks, fixed

**Pull request merged**